haskell-src/Concordium/Types/ProtocolLevelTokens/CBOR.hs

```diff
@@ -647,6 +647,88 @@ encodeTokenHolderTransaction = encodeSequence encodeTokenHolderOperation . token
 
+-- * Token module events
+
+-- | A token-module generated event as part of executing a transaction.
+data EncodedTokenEvent = EncodedTokenEvent
+    { -- | The type of the event. At most 255 bytes.
+      eteType :: !BSS.ShortByteString,
+      -- | (Optional) CBOR-encoded details.
+      eteDetails :: !(Maybe BSS.ShortByteString)
+    }
+    deriving (Eq, Show)
+
+data TokenEvent
+    = -- | An account was added to the allow list.
+      AddAllowListEvent !TokenHolder
+    | -- | An account was removed from the allow list.
+      RemoveAllowListEvent !TokenHolder
+    | -- | An account was added to the deny list.
+      AddDenyListEvent !TokenHolder
+    | -- | An account was removed from the deny list.
+      RemoveDenyListEvent !TokenHolder
+    deriving (Eq, Show)
+
+-- | Encode a 'TokenEvent' as an 'EncodedTokenEvent'.
+encodeTokenEvent :: TokenEvent -> EncodedTokenEvent
+encodeTokenEvent = \case
+    AddAllowListEvent target ->
+        EncodedTokenEvent{eteType = "addAllowList", eteDetails = tokenHolderDetails target}
+    RemoveAllowListEvent target ->
+        EncodedTokenEvent{eteType = "removeAllowList", eteDetails = tokenHolderDetails target}
+    AddDenyListEvent target ->
+        EncodedTokenEvent{eteType = "addDenyList", eteDetails = tokenHolderDetails target}
+    RemoveDenyListEvent target ->
+        EncodedTokenEvent{eteType = "removeDenyList", eteDetails = tokenHolderDetails target}
+  where
+    tokenHolderDetails target =
+        Just . BSS.toShort . CBOR.toStrictByteString $
+            encodeMapLen 1
+                <> encodeString "target"
+                <> encodeTokenHolder target
+
+-- | Decoder for the event details of the list update events.
+--  This is the "token-list-update-details" type in the CDDL schema.
+decodeTokenEventTarget :: Decoder s TokenHolder
+decodeTokenEventTarget = do
+    maybeMapLen <- decodeMapLenOrIndef
+    forM_ maybeMapLen $ \mapLen ->
+        unless (mapLen == 1) $
+            fail $
+                "token-event: expected a map of size 1, but saw " ++ show mapLen
+    label <- decodeString
+    unless (label == "target") $
+        fail $
+            "token-event: expected \"target\" key, but saw "
+                ++ show label
+    target <- decodeTokenHolder
+    when (isNothing maybeMapLen) $ do
+        isEnd <- decodeBreakOr
+        unless isEnd $ fail "token-event: expected end of map"
+    return target
+
+-- | Decode a 'TokenEvent' from an 'EncodedTokenEvent'.
+decodeTokenEvent :: EncodedTokenEvent -> Either String TokenEvent
+decodeTokenEvent EncodedTokenEvent{..} = case eteType of
+    "addAllowList" -> AddAllowListEvent <$> decodeTarget
+    "removeAllowList" -> RemoveAllowListEvent <$> decodeTarget
+    "addDenyList" -> AddDenyListEvent <$> decodeTarget
+    "removeDenyList" -> RemoveDenyListEvent <$> decodeTarget
+    unknownType -> Left $ "token-event: unsupported event type: " ++ show unknownType
```

Comment:
Maybe this is a big discussion to open here, feel free to disregard it if that's the case.

Do we want to gracefully fail in `concordium-base`? It is a publicly available library and I feel like it would make the SDKs (i.e. currently only the rust SDK) job easier not having to translate unions into similar unions but with some variant representing the "unknown".